third_party/move/move-binary-format/src/file_format.rs

```diff
@@ -849,6 +858,116 @@ impl Arbitrary for AbilitySet {
+}
+
+/// The kind of specified access.
+#[derive(Clone, Copy, Eq, Hash, Ord, PartialEq, PartialOrd, Debug)]
+#[cfg_attr(feature = "fuzzing", derive(arbitrary::Arbitrary))]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(proptest_derive::Arbitrary))]
+pub enum AccessKind {
+    Reads,
+    Writes,
+    Acquires, // reads or writes
+}
+
+impl AccessKind {
+    /// Returns true if this access kind subsumes the other.
+    pub fn subsumes(&self, other: &Self) -> bool {
+        use AccessKind::*;
+        match (self, other) {
+            (Acquires, _) => true,
+            (_, Acquires) => false,
+            _ => self == other,
+        }
+    }
+
+    /// Tries to join two kinds, returns None of no intersection.
+    pub fn try_join(self, other: Self) -> Option<Self> {
+        use AccessKind::*;
+        match (self, other) {
+            (Acquires, k) | (k, Acquires) => Some(k),
+            (k1, k2) if k1 == k2 => Some(k1),
+            _ => None,
+        }
+    }
+}
+
+impl fmt::Display for AccessKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use AccessKind::*;
+        match self {
+            Reads => f.write_str("reads"),
+            Writes => f.write_str("writes"),
+            Acquires => f.write_str("acquires"),
+        }
+    }
+}
+
+/// The specification of a resource in an access specifier.
+#[derive(Clone, Eq, Hash, Ord, PartialEq, PartialOrd, Debug)]
+#[cfg_attr(feature = "fuzzing", derive(arbitrary::Arbitrary))]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(proptest_derive::Arbitrary))]
+pub enum ResourceSpecifier {
+    /// Any resource
+    Any,
+    /// A resource declared at the given address.
+    DeclaredAtAddress(AddressIdentifierIndex),
+    /// A resource declared in the given module.
+    DeclaredInModule(ModuleHandleIndex),
+    /// An explicit resource
+    Resource(StructHandleIndex),
+    /// A resource instantiation.
+    ResourceInstantiation(StructHandleIndex, SignatureIndex),
+}
+
+/// The specification of an address in an access specifier.
+#[derive(Clone, Eq, Hash, Ord, PartialEq, PartialOrd, Debug)]
+#[cfg_attr(feature = "fuzzing", derive(arbitrary::Arbitrary))]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(proptest_derive::Arbitrary))]
+pub enum AddressSpecifier {
+    /// Resource can be stored at any address.
+    Any,
+    /// A literal address representation.
+    Literal(AddressIdentifierIndex),
+    /// An address derived from a parameter of the current function.
+    Parameter(
+        /// The index of a parameter of the current function. If `modifier` is not given, the
+        /// parameter must have address type. Otherwise `modifier` must be a function which takes
+        /// a value (or reference) of the parameter type and delivers an address.
+        #[cfg_attr(any(test, feature = "fuzzing"), proptest(strategy = "0u8..63"))]
+        LocalIndex,
+        /// If given, a function applied to the parameter. This is a well-known function which
```

Comment:
What does it mean to be well-known? Will there be a white-list?

Reply:
"well-known" is a notion used in language design for entities which appear like regular user items but have a special meaning to the compiler or runtime. Right now there are only two of those functions, `signer::address_of` and `object::owner`, but here in the file format we do not want to hardcode this since every change would be breaking.